src/hobbits-plugins/importerexporters/UsbDevice/usbdevice.cpp

```diff
@@ -0,0 +1,436 @@
+            return -2;
+            break;
+
+        case LIBUSB_ERROR_ACCESS:
+            return -3;
+            break;
+        
+         case LIBUSB_ERROR_NO_DEVICE:
+            return -4;
+            break;
+        default:
+            return -5;
+            break;
+        }
+    return 0;
+    }
+}
+
+/**
+ * @brief closes all the libusb variables and exits the libusb session after reading all needed data or encountering an error.
+ * 
+ * @param closeDevice  is there a device handle initialized to be able to close, not always true as if there
+ *  is an error initializing a device handle. 
+ */
+void USBDevice::exitLibusb(bool closeDevice){
+    if(closeDevice){ //check if there is a device handle to close
+    libusb_close(m_handle);
+    }
+    libusb_free_config_descriptor(m_config);
+    libusb_exit(m_ctx);
+}
+
+/**
+ * @brief The function to actually read the binary data from the USB device, imports parameters, determines the transfer needed,
+ * checks for device handle errors, detaches the kernel driver, reads the data, resets the device, updates the progressbar, 
+ * checks if action is cancelled and then waits for the duration of transfer delay, and repeats for the amount in NumTransfers
+ * 
+ * @param parameters the parameters from the importer UI needed to set libusb to read the data form the proper device and endpoint
+ * @param progress the pointer resposible for updating the progressbar and checking for cancelled tasks
+ * @return QSharedPointer<ImportResult> - the pointer towards the bitcontainer so that hobbits can read the data being imported
+ */
+QSharedPointer<ImportResult> USBDevice::importBits(const Parameters &parameters,
+                                                      QSharedPointer<PluginActionProgress> progress)
+{
+    //validating params
+    QStringList invalidations = m_importDelegate->validate(parameters);
+    if (!invalidations.isEmpty()) {
+        return ImportResult::error(QString("Invalid parameters passed to %1:\n%2").arg(name()).arg(invalidations.join("\n")));
+    }
+
+    //getting parameters
+    m_deviceNum = parameters.value("DeviceNum").toInt();
+    m_interfaceNum = parameters.value("InterfaceNum").toInt();
+    m_altSetNum = parameters.value("AltSetNum").toInt();
+    m_endpointNum = parameters.value("EndpointNum").toInt();
+    int transferNum = parameters.value("TransferNum").toInt();
+    int transferDelay = parameters.value("TransferDelay").toInt();
+    int transferTimeout = parameters.value("TransferTimeout").toInt();
+    int transferType = parameters.value("TransferType").toInt();
+    int transferSize = parameters.value("TransferSize").toInt();
+    QString transferTypeStr = QString::number(transferType);
+    QString deviceName = parameters.value("DeviceName").toString();
+    bool includeTimeout = parameters.value("TimeoutIn").toBool();
+    QSharedPointer<RangeSequence> frames = RangeSequence::createEmpty();
+    QByteArray largeBuffer;
```

Comment:
Honestly I think the current QByteArray is perfectly fine as it covers 99.9% of possible use cases for the plugin. The largest possible transfer over usb is a max of 1 KiB and so a user would have to do a million transfers to get to a GiB with the largest of bulk endpoints. To be fair this is a possibility, but I don't expect it to be a major use case for most users , and so honestly I don't think its entirely necessary to adopt because of that. BitArray also complicates the appending of data from the small buffer to the large buffer, as there is no one defined method, and instead you have to keep track of the total data transferred, it becomes a lot of hassle for not a lot of benefit resulting from it.